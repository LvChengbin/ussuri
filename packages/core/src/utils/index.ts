/******************************************************************
 * Copyright (C) 2022-2023 NextSeason
 *
 * @File: utils/index.ts
 *
 * This file is part of NextSeason projects.
 * Code in this file can not be copied and/or distributed without the
 * express permission of NextSeason. Inc
 ******************************************************************/

export * from './change-case';
export * from './compose';
export * from './create-constructor-interceptors';
export * from './escape-regexp';
export * from './join-pattern';
export * from './merge-application-options';
export * from './pattern-combinations';
