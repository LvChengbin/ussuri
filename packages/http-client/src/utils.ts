/******************************************************************
 * Copyright (C) 2022 LvChengbin
 *
 * File: src/utils.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 05/10/2022
 * Description:
 ******************************************************************/

import { RequestOptions } from './interfaces';

export function joinPath( ...pieces: string[] ): string {
    const path = pieces.map( x => x.replace( /^\/+|\/+$/g, '' ) ).join( '/' );
    if( !path.startsWith( '/' ) ) return '/' + path;
    return path;
}

const REQUET_METHODS_WITH_PAYLOAD = [ 'POST', 'PUT', 'PATCH' ];

export function isRequestWithPayload( method: string ): boolean {
    return REQUET_METHODS_WITH_PAYLOAD.includes( method.toUpperCase() );
}

export function isAbsoluteURL( url: string ): boolean {
    return /^([a-z][a-z\d+\-.]*:)?\/\//i.test( url );
}

export function buildURL( base: string | undefined, url: string ): string {
    if( !base && !url ) return '';
    if( !base || isAbsoluteURL( url ) ) return url;
    return base.replace( /\/+$/, '' ) + '/' + url.replace( /^\/+/, '' );
}

export function interpolateURL( url: string, data: Record<string, any> ): string {
    return url.replace( /:([a-zA-Z_][a-zA-Z0-9_]*)/g, ( m: string, n: string ) => {
        const value = data[ n ];
        if( typeof value === 'string' || typeof value === 'number' ) return String( value );
        // eslint-disable-next-line no-console
        console.warn( `Invalid url param ${value} for ${url}` );
        return '';
    } );
}

export function spreadFormData<T extends Record<string, any>>( formdata: T ): Record<string, any> {
    if( !( formdata instanceof FormData ) ) return formdata;
    const output: Record<string, any> = {};
    for( const [ key, value ] of formdata.entries() ) {
        output[ key ] = value;
    }
    return output;
}

export function mergeData( a: FormData, b: FormData | Record<string, any> ): FormData;
export function mergeData( a: FormData | Record<string, any>, b: FormData ): FormData;
export function mergeData<T1 extends Record<string, any>, T2 extends Record<string, any>>( a: T1, b: T2 ): T1 & T2;
export function mergeData( a: FormData | Record<string, any>, b: FormData | Record<string, any> ): FormData | Record<string, any> {

    const data = {
        ...spreadFormData( a ),
        ...spreadFormData( b )
    } as Record<string, any>;

    if( a instanceof FormData || b instanceof FormData ) {
        const formdata = new FormData();
        Object.keys( data ).forEach( ( key: string ) => {
            formdata.append( key, data[ key ] );
        } );
        return formdata;
    }

    return data;
}

export function mergeOptions( request: RequestOptions, base: RequestOptions ): RequestOptions {
    const headers = { ...base.headers, ...request.headers };
    const params = { ...base.params, ...request.params };
    const query = { ...base.query, ...request.query };
    const payload = mergeData( base.payload, request.payload );
    const data = mergeData( base.data, request.data );

    const url = buildURL(
        typeof base.service === 'string'
            ? buildURL( base.service, base.url ?? '' )
            : base.url,
        request.url ?? ''
    );

    return {
        ...base, ...request,
        headers, params, payload, data, query, url
    };
}

export function lowerCaseFirstLetter( str: string ): string {
    return str.charAt( 0 ).toLowerCase() + str.slice( 1 );
}
