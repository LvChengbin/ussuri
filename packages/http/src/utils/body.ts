/******************************************************************
 * Copyright (C) 2020 LvChengbin
 *
 * File: src/body-parser.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 12/09/2020
 * Description:
 ******************************************************************/

import qs from 'qs';
import bytes from 'bytes';
import formidable, { Fields, Files, Options as MultipartOptions, BufferEncoding } from 'formidable';
import cobody, { Options as CobodyOptions } from 'co-body';
import { Context } from '../context';

export { Files as RequestFiles, File as RequestFile } from 'formidable';

/**
 * The type of the return value of `cobody` while `options.returnRawBody` is set to `true`.
 * The `parsed` part will be in different types while using different request types.
 *  - it will be `string` while `Content-Type` matches `text`.
 *  - it will be `ReturnType<typeof qs.parse>` while `Content-Type` matches `form`.
 *  - it will be `ReturnType<typeof JSON.parse>` while `Content-Type` matches `json`.
 */
type RawBody = {
    parsed: string | ReturnType<typeof qs.parse> | ReturnType<typeof JSON.parse>;
    raw: string;
};

export interface MultipartBody {
    fields: Fields;
    files: Files ;
}

async function parseMultipart( context: Context, options: MultipartOptions ): Promise<MultipartBody> {
    return new Promise( ( resolve, reject ) => {
        const form = formidable( options );

        form.parse( context.req!, ( err, fields, files ) => {
            if( err ) reject( err );
            else resolve( { fields, files } );
        } );
    } );
}

/**
 * BodyParseOptions extends Cobody.Options.
 *
 * @see https://github.com/cojs/co-body
 * @see https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/co-body/index.d.ts
 */
export interface BodyParseOptions extends CobodyOptions {
    encoding?: BufferEncoding;
    limit?: string;
    strict?: boolean;
    queryString?: qs.IParseOptions;
    returnRawBody?: boolean;
    jsonTypes?: string[];
    formTypes?: string[];
    textTypes?: string[];
    multipartOptions?: MultipartOptions;
}

/**
 * get body data by setting `options.returnRawBody` to `true`, it will return the {@link RawBody} only when `Content-Type` doesn't match `multipart`.
 *
 * @param context - the context object
 * @param options - the options object. {@link BodyParseOptions}
 *
 * @returns the parsed body with `fields` and `files`.
 */
export async function bodyParse(
    context: Context,
    options: Readonly<BodyParseOptions> = {}
): Promise<RawBody | MultipartBody | ReturnType<typeof cobody> | null> {

    if( !context.req || context.request.type === '' ) return null;

    const encoding = options.encoding ?? 'utf-8';
    const limit = '20mb';

    if( context.request.is( 'multipart' ) ) {

        const { multipartOptions = {} } = options;
        const maxSizeOptions: {
            maxFieldsSize?: number;
            maxFileSize?: number;
        } = {};

        if( typeof multipartOptions.maxFileSize === 'string' ) {
            maxSizeOptions.maxFileSize = bytes.parse( multipartOptions.maxFileSize );
        }

        if( typeof multipartOptions.maxFieldsSize === 'string' ) {
            maxSizeOptions.maxFieldsSize = bytes.parse( multipartOptions.maxFieldsSize );
        }

        return parseMultipart( context, {
            encoding,
            multiples : true,
            minFileSize : 1,
            maxFileSize : 209715200, // 200mb
            maxFields : 2000,
            maxFieldsSize : 20971520, // 20mb
            ...multipartOptions,
            ...maxSizeOptions
        } );
    }

    return cobody( context.req, { encoding, limit, returnRawBody : false, ...options } );
}
