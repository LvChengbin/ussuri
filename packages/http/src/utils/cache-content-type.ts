/******************************************************************
 * Copyright (C) 2021 LvChengbin
 *
 * File: utils/cache-content-type.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 07/31/2021
 * Description:
 ******************************************************************/

import mimetypes from 'mime-types';
import { LRUCache } from 'lru-cache';

const cache = new LRUCache<string, string | false>( { max : 1000 } );

export default ( type: string ): string | false => {
    let mimetype = cache.get( type );
    if( !mimetype ) {
        mimetype = mimetypes.contentType( type );
        cache.set( type, mimetype );
    }
    return mimetype;
};
