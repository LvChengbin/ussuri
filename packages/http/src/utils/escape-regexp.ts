/******************************************************************
 * Copyright (C) 2022 LvChengbin
 *
 * File: utils/escape-regexp.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 07/02/2022
 * Description:
 ******************************************************************/

export function escapeRegexp( s: string ): string {
    return s.replace( /[/|\\{}()[\]^$+*?.]/g, '\\$&' ).replace( /-/g, '\\x2d' );
}
