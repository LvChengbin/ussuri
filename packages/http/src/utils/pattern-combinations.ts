/******************************************************************
 * Copyright (C) 2022 LvChengbin
 *
 * File: utils/pattern-combinations.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 03/15/2022
 * Description:
 ******************************************************************/

export function patternCombinations<T extends string | RegExp>( ...patterns: ( T[] | T )[] ): T[][] {
    let combinations: T[][] = [];

    patterns.forEach( ( group: T[] | T ) => {

        if( !Array.isArray( group ) ) group = [ group ];

        if( combinations.length === 0 ) {
            combinations = ( group as T[] ).map( x => [ x ] );
            return;
        }

        const tmp: T[][] = [];

        combinations.forEach( ( item: T[] ) => {
            ( group as T[] ).forEach( ( pattern: T ) => {
                tmp.push( [ ...item, pattern ] );
            } );
        } );

        combinations = tmp;

    } );

    return combinations;
}
