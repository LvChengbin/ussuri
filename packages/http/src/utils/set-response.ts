/******************************************************************
 * Copyright (C) 2022 LvChengbin
 *
 * File: utils/set-response.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 04/12/2022
 * Description:
 ******************************************************************/

import { Context } from '../context';

export function setResponse<T extends Context>( response: unknown, context: T ): T {
    if( response instanceof Context ) return response as T;
    context.body = typeof response === 'number' ? String( response ) : response;
    return context;
}
