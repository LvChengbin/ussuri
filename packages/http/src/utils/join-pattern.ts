/******************************************************************
 * Copyright (C) 2022 LvChengbin
 *
 * File: utils/join-pattern.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 03/14/2022
 * Description:
 ******************************************************************/

import { pathToRegexp, Key } from 'path-to-regexp';
import { escapeRegexp as escape } from './escape-regexp';

interface JoinPatternResult {
    pattern: RegExp;
    keys: ( string | number )[];
}

interface JoinPatternFunction {
    ( ...args: ( string | RegExp )[] ): JoinPatternResult;
}

export interface JoinPatternOptions {
    start?: string;
    end?: string;
    delimiter?: string;
    separator?: string;
}

export function joinPattern( options: JoinPatternOptions = {} ): JoinPatternFunction {

    const { start = '/', delimiter = '/#?', separator = '/' } = options;

    const trimStartRe = new RegExp( `^${escape( escape( start ) )}` );

    /**
     * create regexp segment like `[\/#?]`.
     */
    const delimiterReSeg = '[' + escape( delimiter ?? '/#?' ) + ']';
    const separatorReSeg = escape( separator );

    const trimPathPatternRe = new RegExp( `(^${separatorReSeg}+)|(${separatorReSeg}+$)`, 'g' );

    /**
     * Remove characters in delimiter of path pattern
     *
     * `/account/:id` -> `account/:id`
     */
    const trimPathPattern = ( str: string ): string => str.replace( trimPathPatternRe, '' );

    const trimReSourceRe = new RegExp( `(^\\^)|((?:${escape( delimiterReSeg )}\\?)\\$?$)`, 'g' );

    const trimReSource = ( s: string ): string => s.replace( trimReSourceRe, '' );

    return ( ...patterns: ( string | RegExp )[] ): JoinPatternResult => {

        const flags: string[] = [];
        const sources: string[] = [];
        const keys: Key[] = [];

        patterns.forEach( ( pattern: string | RegExp ) => {

            const k: Key[] = [];

            if( typeof pattern === 'string' ) {
                pattern = pathToRegexp( trimPathPattern( pattern ), k, { delimiter } );
            } else {
                /**
                 * path-to-regex can collect named group in regexp
                 */
                pattern = pathToRegexp( pattern, k, { delimiter } );
            }

            keys.push( ...k );

            const trimedSource = trimReSource( ( pattern as RegExp ).source );

            if( trimedSource.length ) {
                sources.push( trimedSource );
                flags.push( ...( pattern as RegExp ).flags ?? '' );
            }

        } );

        const source = sources.join( separator ).replace( trimStartRe, '' );

        return {
            pattern : new RegExp(
                `^${start}${source}${ source ? delimiterReSeg : '' }?$`,
                [ ...new Set( flags ) ].join( '' )
            ),
            keys : keys.map( ( key: Key ) => key.name )
        };
    };
}
