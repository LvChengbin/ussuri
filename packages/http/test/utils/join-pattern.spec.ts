/******************************************************************
 * Copyright (C) 2022 LvChengbin
 *
 * File: utils/join-pattern.spec.ts
 * Author: LvChengbin<lvchengbin59@gmail.com>
 * Time: 03/14/2022
 * Description:
 ******************************************************************/

import { joinPattern } from '../../src/utils/join-pattern';

function getMatches( pattern: RegExp, path: string ): string[] | false {
    let matches = pattern.exec( path );

    if( !matches ) return false;

    matches = Array.prototype.slice.call( pattern.exec( path ) );
    matches.shift();

    return matches;
}

describe( 'utils/join-pattern', () => {

    const join = joinPattern( '/' );

    it( 'empty path', () => {
        const { keys, pattern } = join( '' );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [] );
        const matches = getMatches( pattern, '/' );
        expect( matches ).toEqual( [] );
        expect( getMatches( pattern, '//' ) ).toBeFalsy();
    } );

    it( 'multiple empty path', () => {
        const { keys, pattern } = join( '', '' );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [] );
        const matches = getMatches( pattern, '/' );
        expect( matches ).toEqual( [] );
        expect( getMatches( pattern, '//' ) ).toBeFalsy();
    } );

    it( 'single string', () => {
        const { keys, pattern } = join( '/account/:id(\\d+)/:item' );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item' ] );
        const matches = getMatches( pattern, '/account/123/profile' );
        expect( matches ).toEqual( [ '123', 'profile' ] );
        expect( getMatches( pattern, '/account/abc/profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123//profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123#profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );

    it( 'multiple strings', () => {
        const { keys, pattern } = join( '/account/:id(\\d+)', ':item' );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item' ] );
        const matches = getMatches( pattern, '/account/123/profile' );
        expect( matches ).toEqual( [ '123', 'profile' ] );
        expect( getMatches( pattern, '/account/abc/profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123//profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123#profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );

    it( 'should trim separators', () => {
        const { keys, pattern } = join( '/account/:id(\\d+)/', ':item' );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item' ] );
        const matches = getMatches( pattern, '/account/123/profile' );
        expect( matches ).toEqual( [ '123', 'profile' ] );
        expect( getMatches( pattern, '/account/abc/profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123//profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );

    it( 'single regexp', () => {
        const { keys, pattern } = join( /^account\/(?<id>\d+)\/(?<item>[^/]+)/ );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item' ] );
        const matches = getMatches( pattern, '/account/123/profile' );
        expect( matches ).toEqual( [ '123', 'profile' ] );
        expect( getMatches( pattern, '/account/abc/profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );

    it( 'single regexp with ahead "/"', () => {
        const { keys, pattern } = join( /^\/account\/(?<id>\d+)\/(?<item>[^/]+)/ );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item' ] );
        const matches = getMatches( pattern, '/account/123/profile' );
        expect( matches ).toEqual( [ '123', 'profile' ] );
        expect( getMatches( pattern, '/account/abc/profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );

    it( 'multiple regexps', () => {
        const { keys, pattern } = join( /^\/account\/(?<id>\d+)/, /(?<item>[^/]+)/ );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item' ] );
        const matches = getMatches( pattern, '/account/123/profile' );
        expect( matches ).toEqual( [ '123', 'profile' ] );
        expect( getMatches( pattern, '/account/abc/profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123//profile' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );

    it( 'mix string patterns and regexps', () => {
        const { keys, pattern } = join( '/account/:id(\\d+)/', /(?<item>[^/]+)/, ':name' );
        expect( pattern ).toMatchSnapshot();
        expect( keys ).toEqual( [ 'id', 'item', 'name' ] );
        const matches = getMatches( pattern, '/account/123/profile/x' );
        expect( matches ).toEqual( [ '123', 'profile', 'x' ] );
        expect( getMatches( pattern, '/account/abc/profile/x' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/123//profile/x' ) ).toBeFalsy();
        expect( getMatches( pattern, '/account/122' ) ).toBeFalsy();
    } );
} );
