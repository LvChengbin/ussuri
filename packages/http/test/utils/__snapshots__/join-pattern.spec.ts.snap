// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils/join-pattern empty path 1`] = `/\\^\\\\/\\?\\$/`;

exports[`utils/join-pattern mix string patterns and regexps 1`] = `/\\^\\\\/account\\(\\?:\\\\/\\(\\\\d\\+\\)\\)\\\\/\\(\\?<item>\\[\\^/\\]\\+\\)\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\[\\\\/#\\\\\\?\\]\\?\\$/i`;

exports[`utils/join-pattern multiple empty path 1`] = `/\\^\\\\/\\?\\$/`;

exports[`utils/join-pattern multiple regexps 1`] = `/\\^\\\\/account\\\\/\\(\\?<id>\\\\d\\+\\)\\\\/\\(\\?<item>\\[\\^/\\]\\+\\)\\[\\\\/#\\\\\\?\\]\\?\\$/`;

exports[`utils/join-pattern multiple strings 1`] = `/\\^\\\\/account\\(\\?:\\\\/\\(\\\\d\\+\\)\\)\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\[\\\\/#\\\\\\?\\]\\?\\$/i`;

exports[`utils/join-pattern should trim separators 1`] = `/\\^\\\\/account\\(\\?:\\\\/\\(\\\\d\\+\\)\\)\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\[\\\\/#\\\\\\?\\]\\?\\$/i`;

exports[`utils/join-pattern single regexp 1`] = `/\\^\\\\/account\\\\/\\(\\?<id>\\\\d\\+\\)\\\\/\\(\\?<item>\\[\\^/\\]\\+\\)\\[\\\\/#\\\\\\?\\]\\?\\$/`;

exports[`utils/join-pattern single regexp with ahead "/" 1`] = `/\\^\\\\/account\\\\/\\(\\?<id>\\\\d\\+\\)\\\\/\\(\\?<item>\\[\\^/\\]\\+\\)\\[\\\\/#\\\\\\?\\]\\?\\$/`;

exports[`utils/join-pattern single string 1`] = `/\\^\\\\/account\\(\\?:\\\\/\\(\\\\d\\+\\)\\)\\(\\?:\\\\/\\(\\[\\^\\\\/#\\\\\\?\\]\\+\\?\\)\\)\\[\\\\/#\\\\\\?\\]\\?\\$/i`;
